File: javascript-solutions/leetcode/p34-firstLastPosition.js
/// # 34. Find First and Last Position of Element in Sorted Array
/// Given an array of integers `nums` sorted in ascending order, find the starting and ending
/// position of a given `target` value.
/// If target is not found in the array, return `[-1, -1]`.
/// You must write an algorithm with `O(log n)` runtime complexity.


var searchRange = function(nums, target) {

  function findRightmostIndex(arr, startIndex) {
    const target = arr[startIndex];
    let start = startIndex;
    let end = arr.length;

    while (start < end) {
      if (arr[end-1] === target) {
        return end - 1;
      }
      const mid = Math.floor((start + end) / 2);
      if (arr[mid] === target) {
        start = mid;
      } else {
        // arr[mid] > target, target is to left of mid
        end = mid;
      }
    }
    return start;
  }

  function findLeftmostIndex(arr, endIndex) {
    const target = arr[endIndex];
    let end = endIndex;
    let start = 0;

    while (start <= end) {
      if (arr[start] === target) {
        return start;
      }
      const mid = Math.floor((start + end) / 2);
      if (arr[mid] === target) {
        end = mid;
      } else {
        // arr[mid] < target, so target is to right of mid
        start = mid + 1;
      }
    }
  }
  
  let start = 0;
  let end = nums.length;

  while (start < end) {

    // base case, we have an array of length 1
    if (nums.length === 1) {
      if (nums[0] === target) {
        return [0,0];
      } else {
        return [-1, -1];
      }
    }

    // check if either nums[first] or nums[last] is == target
    if (nums[start] === target) {
      const sidx = findLeftmostIndex(nums, start);
      const eidx = findRightmostIndex(nums, start);
      return [sidx, eidx];
    }

    if (nums[end - 1] === target) {
      const sidx = findLeftmostIndex(nums, end - 1);
      const eidx = findRightmostIndex(nums, end - 1);
      return [sidx, eidx];
    }

    // otherwise check midpoint index
    const mid = Math.floor((start + end) / 2);
    if (nums[mid] === target) {
      let sidx = findLeftmostIndex(nums, mid);
      let eidx = findRightmostIndex(nums, mid);
      return [sidx, eidx];
    }

    if (target <= nums[mid]) {
      // target is on left side
      end = mid;
    } else {
      // target is on right side
      start = mid + 1;
    }
  }

  return [-1, -1];
};

const nums =  [5,7,7,8,8,10];
let res = searchRange(nums, 8);
console.log(res);

const nums2 =  [5,7,7,8,8,10];
let res2 = searchRange(nums2, 6);
console.log(res2);

File: rust-solutions/leetcode/src/bin/p34-first-last-position.rs
use std::cmp::Ordering;

/// # 34. Find First and Last Position of Element in Sorted Array
/// Given an array of integers `nums` sorted in ascending order, find the starting and ending
/// position of a given `target` value.
/// If target is not found in the array, return `[-1, -1]`.
/// You must write an algorithm with `O(log n)` runtime complexity.

pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
    // returns rightmost index(inclusive) of some target element within nums, beginning at start inclusive,
    // it's assumed that nums[start] will initially contain the target element you are searching for.
    fn find_rightmost_index(nums: &[i32], start: usize) -> usize {
        let target = nums[start];
        let mut start = start;
        let mut end = nums.len();
        while start < end {
            if nums[end - 1] == target {
                return end - 1;
            }
            let mid = (start + end) / 2;
            match nums[mid].cmp(&target) {
                Ordering::Equal => start = mid,
                Ordering::Greater => end = mid,
                _ => (),
            }
        }
        start
    }

    // returns left-most index(inclusive) of some target element within nums, beginning at end inclusive,
    // it's assumed that nums[end] will initially contain the target element you are searching for.
    fn find_leftmost_index(nums: &[i32], end: usize) -> usize {
        let target = nums[end];
        let mut end = end;
        let mut start = 0;
        while start <= end {
            if nums[start] == target {
                return start;
            }
            let mid = (start + end) / 2;
            match nums[mid].cmp(&target) {
                Ordering::Equal => end = mid,
                Ordering::Less => start = mid + 1,
                _ => (),
            }
        }
        end
    }

    // indices into the current nums slice we are examining
    let mut start = 0;
    let mut end = nums.len();

    while start < end {
        // base case, array of one element
        match nums.len() {
            1 if nums[0] == target => return vec![0, 0],
            1 if nums[0] != target => return vec![-1, -1],
            _ => (), // continue
        }

        // check if either nums[first] or nums[last] is == target
        if nums[start] == target {
            let si = find_leftmost_index(&nums, start);
            let ei = find_rightmost_index(&nums, start);
            return vec![si as i32, ei as i32];
        }

        if nums[end - 1] == target {
            let si = find_leftmost_index(&nums, end - 1);
            let ei = find_rightmost_index(&nums, end - 1);
            return vec![si as i32, ei as i32];
        }

        // now check nums[mid] where mid is the middle index between start and end
        let mid = (start + end) / 2;
        if nums[mid] == target {
            let si = find_leftmost_index(&nums, mid);
            let ei = find_rightmost_index(&nums, mid);
            return vec![si as i32, ei as i32];
        }

        if target <= nums[mid] {
            // target is on left side
            end = mid;
        } else {
            // target is on right side
            start = mid + 1;
        }
    }

    vec![-1, -1]
}

fn main() {}

#[cfg(test)]
mod tests {
    use crate::search_range;

    #[test]
    fn example1() {
        let nums = vec![5, 7, 7, 8, 8, 10];
        let target = 8;
        assert_eq!(search_range(nums, target), vec![3, 4])
    }

    #[test]
    fn example2() {
        let nums = vec![];
        let target = 0;
        assert_eq!(search_range(nums, target), vec![-1, -1])
    }

    #[test]
    fn example3() {
        let nums = vec![8, 8, 8, 8];
        let target = 8;
        assert_eq!(search_range(nums, target), vec![0, 3])
    }

    #[test]
    fn example4() {
        let nums = vec![8];
        let target = 8;
        assert_eq!(search_range(nums, target), vec![0, 0])
    }

    #[test]
    fn example4_1() {
        let nums = vec![8];
        let target = 6;
        assert_eq!(search_range(nums, target), vec![-1, -1]);
    }

    #[test]
    fn example4_2() {
        let nums = vec![0, 8];
        let target = 8;
        assert_eq!(search_range(nums, target), vec![1, 1]);
    }

    #[test]
    fn example4_3() {
        let nums = vec![7, 8];
        let target = 9;
        assert_eq!(search_range(nums, target), vec![-1, -1]);
    }

    #[test]
    fn example4_4() {
        let nums = vec![2, 2];
        let target = 1;
        assert_eq!(search_range(nums, target), vec![-1, -1]);
    }

    #[test]
    fn example4_5() {
        let nums = vec![2, 2];
        let target = 3;
        assert_eq!(search_range(nums, target), vec![-1, -1]);
    }

    #[test]
    fn example5() {
        let nums = vec![1, 2, 3, 4, 5, 6];
        let target = 8;
        assert_eq!(search_range(nums, target), vec![-1, -1])
    }

    #[test]
    fn example5_5() {
        let nums = vec![5, 7, 7, 8, 8, 10];
        let target = 6;
        assert_eq!(search_range(nums, target), vec![-1, -1])
    }

    #[test]
    fn example6() {
        let nums = vec![5, 6, 7, 8, 9, 10, 11];
        let target = 8;
        assert_eq!(search_range(nums, target), vec![3, 3])
    }
}
